contracts/user/ConsumerWrapper.sol

```diff
@@ -0,0 +1,232 @@
+                isFlatFeePromotionEnabledPermanently,
+                flatFeePromotionStartTimestamp,
+                flatFeePromotionEndTimestamp
+            );
+        }
+    }
+
+    function _calculateTierFee(uint64 reqCount, uint256 lastRequestTimestamp, uint64 reqCountInCurrentPeriod)
+        internal
+        view
+        returns (uint32 tierFee)
+    {
+        // Use the new struct here.
+        FeeConfig memory feeConfig = _getFlatFeeConfig();
+        uint64 reqCountCalc;
+        if (feeConfig.isFlatFeePromotionEnabledPermanently) {
+            reqCountCalc = reqCount;
+        } else if (
+            feeConfig
+                //solhint-disable-next-line not-rely-on-time
+                .flatFeePromotionStartTimestamp <= block.timestamp
+            //solhint-disable-next-line not-rely-on-time
+            && block.timestamp <= feeConfig.flatFeePromotionEndTimestamp
+        ) {
+            if (lastRequestTimestamp < feeConfig.flatFeePromotionStartTimestamp) {
+                reqCountCalc = 1;
+            } else {
+                reqCountCalc = reqCountInCurrentPeriod + 1;
+            }
+        }
+        return IAdapter(adapter).getFeeTier(reqCountCalc) * feeConfig.flatFeePromotionGlobalPercentage / 100;
+    }
+
+    function getRandomNumber(
+        uint64 subId,
+        bytes32 entityId,
+        uint32 gasLimit,
+        address callbackAddress,
+        bytes4 callbackFunctionSelector)
+        external payable returns (bytes32 requestId)
+    {
+        if (gasLimit > _MAX_GAS_LIMIT) {
+            revert GasLimitTooBig(gasLimit, _MAX_GAS_LIMIT);
+        }
+        subId = _fundSubId(subId, gasLimit);
+        (uint16 requestConfirmations,,,,,,) = IAdapter(adapter).getAdapterConfig();
+        bytes memory params;
+        requestId = _rawRequestRandomness(
+            RequestType.Randomness, params, subId, 0, requestConfirmations, gasLimit, tx.gasprice * 3
+        );
+        pendingRequests[requestId] = RequestData({
+            entityId: entityId,
+            callbackAddress: callbackAddress,
+            callbackFunctionSelector: callbackFunctionSelector // Store the selector in the request data
+        });
+        emit RandomNumberRequest(callbackAddress, requestId, entityId);
+    }
+
+    // Modify the _fulfillRandomness function to use the stored callback function selector
+    function _fulfillRandomness(bytes32 requestId, uint256 randomness) internal override {
+        RequestData memory requestData = pendingRequests[requestId];
+        // Using the callbackFunctionSelector to call the function on the requesting contract
+        (bool success,) = requestData.callbackAddress.call(
+            abi.encodeWithSelector(
+                requestData.callbackFunctionSelector,
+                requestId,
+                randomness,
+                requestData.entityId
+            )
+        );
+        if (!success) {
+            revert CallbackFailed(requestId);
```

Comment:
The successful execution of callback function is not guaranteed by us. If callback fails, I suggest we record it as a field of `RandomNumberResult` event and compensate working nodes normally.